csharp/src/Ice/Connection.cs

```diff
@@ -1037,10 +1037,21 @@ private async ValueTask InvokeAsync(Current current, IncomingRequestFrame reques
                         var responseFrame = new IncomingResponseFrame(_communicator,
                             readBuffer.Slice(Ice1Definitions.HeaderSize + 4));
                         ProtocolTrace.TraceFrame(_communicator, readBuffer, responseFrame);
-                        if (_requests.TryGetValue(requestId, out TaskCompletionSource<IncomingResponseFrame>? response))
+                        if (_requests.TryGetValue(requestId,
+                                out TaskCompletionSource<IncomingResponseFrame>? responseTaskCompletionSource))
                         {
                             _requests.Remove(requestId);
-                            response.SetResult(responseFrame);
+                            // We can't call SetResult directly from here as it might be trigger the continuations
+                            // to run synchronously and it wouldn't be safe to run a continuation with the mutex
+                            // locked.
+                            //
+                            // TODO: Running continuations of synchronous call would be safe but we don't know
+                            // here if this is a synchronous call or not. We could if we provided this information
+                            // to SendRequestAsync and kept track.
+                            incoming = () => {
+                                responseTaskCompletionSource.SetResult(responseFrame);
+                                return new ValueTask(Task.CompletedTask);
```

Comment:
This lambda expression that always returns a completed task looks odd. Presumably "incoming" typically returns a live task?

Reply:
Yes, see the code that parses incoming request few lines below:
```
                                adapter = _adapter;
                                int requestId = InputStream.ReadInt(readBuffer.AsSpan(Ice1Definitions.HeaderSize, 4));
                                var current = new Current(_adapter, request, requestId, this);
                                incoming = () => InvokeAsync(current, request, compressionStatus);
                                ++_dispatchCount;
```